internal/auditd/dirreader.go

```diff
@@ -337,37 +340,135 @@ func (o *osFileSystem) Open(filePath string) (io.ReadSeekCloser, error) {
 	f, err := fsi.Open(filePath)
 	if err != nil {
 		return 0, err
 	}
 	defer f.Close()
 
-	return readLines(ctx, f, l)
-}
-
-// readLines reads data from reader until EOF using bufio.ScanLines.
-// It writes each line to the lines chan and returns the number of
-// bytes read from reader.
-func readLines(ctx context.Context, reader io.Reader, lines chan<- string) (int64, error) {
-	scanner := bufio.NewScanner(reader)
+	scanner := bufio.NewScanner(f)
 	var numBytesRead int64
 
 	for scanner.Scan() {
 		line := scanner.Text()
-		numBytesRead += int64(len(line) + 1)
+		numBytesRead += int64(len(line)) + 1
 
 		select {
 		case <-ctx.Done():
 			return numBytesRead, ctx.Err()
-		case lines <- line:
+		case l <- line:
 			// continue.
 		}
 	}
 
 	return numBytesRead, scanner.Err()
 }
+
+// readLines reads from reader and writes each line to the lines chan.
+// It then returns the total number of bytes read in terms of lines
+// processed (i.e., if it reads 100 bytes, but only 90 of those bytes
+// were lines ending with a newline, it returns 90). Buffered data is
+// not counted in the byte count.
+//
+// The supplied backoffSleeper is used to gradually backoff reads that
+// result in an io.EOF. The assumption is that the caller will execute
+// this function again when more data is available to read. Making such
+// a determination is left up to the caller.
+//
+// An io.EOF error is not returned to the caller.
+func readLines(ctx context.Context, reader io.Reader, lines chan<- string, backoff *backoffSleeper) (int64, error) {
+	bufioReader := bufio.NewReader(reader)
+	var numBytesRead int64
+
+	for {
+		lineRaw, err := bufioReader.ReadString('\n')
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Receiving an io.EOF means that no newline
+				// character was found. Instead of returning
+				// that to the caller, we can use a backoff
+				// coupled with returning a nil error to
+				// ultimately allow the caller to try again.
+				err = backoff.incAndSleep(ctx)
+				if err != nil {
+					return numBytesRead, err
+				}
+
+				return numBytesRead, nil
+			}
+
+			return numBytesRead, err
+		}
+
+		backoff.dec()
+
+		lineLen := len(lineRaw)
+
+		numBytesRead += int64(lineLen)
+
+		var line string
+		if lineLen > 0 {
+			// Remove the trailing newline character.
+			//
+			// Yes, this can result in an empty string.
+			// Allowing an empty string maintains
+			// backwards compatibility with our
+			// existing unit tests.
+			line = lineRaw[0 : lineLen-1]
+		}
+
+		select {
+		case <-ctx.Done():
+			return numBytesRead, ctx.Err()
+		case lines <- line:
+			// continue.
+		}
+	}
+}
+
+// backoffSleeper helps implement a backoff timer that results in
+// an incremental sleep. The increments are focused towards file
+// IO. Being completely honest, they are a SWAG at best.
+type backoffSleeper struct {
+	sleepFor time.Duration
+}
```

Comment:
Instead of implementing your own backoff mechanism, why not use https://pkg.go.dev/github.com/cenkalti/backoff/v4 ?

Reply:
It had not occurred to me to do that - mainly because we wanted a fast turn-around on this.

The backoff timing part feels less critical to me than handling `io.EOF` properly (prior to this change, my use of `bufio.Scanner` did not account for this). I did not know about that library either, which at least appears to have zero external dependencies and is based on an existing implementation that (presumably) works.